### PR TITLE
Implement recent window.open spec changes, and mostly end up stuck in random LibURL things

### PR DESCRIPTION
**Pull request by AtkinsSJ**
There may be some WPT tests affected by this, I'm honestly not sure. This started as "apply this small spec change" and became a real yak-shaving party of unimplemented things relying on other unimplemented things. (And "X would be so much nicer if I did Y..." scenarios.) I don't know how observable these changes even are. 😅

### CODE REVIEW

File: Libraries/LibWeb/HTML/WorkerLocation.cpp [hunk 1]
```diff
@@ -60,8 +60,6 @@ WebIDL::ExceptionOr<String> WorkerLocation::host() const
 // https://html.spec.whatwg.org/multipage/workers.html#dom-workerlocation-hostname
 WebIDL::ExceptionOr<String> WorkerLocation::hostname() const
```
Comment (shannonbooth) [inline Libraries/LibWeb/HTML/WorkerLocation.cpp:61]: We can remove the ExceptionOr now :^)

File: Libraries/LibWeb/HTML/HTMLHyperlinkElementUtils.cpp [hunk 2]
```diff
@@ -206,7 +206,8 @@ String HTMLHyperlinkElementUtils::hostname() const
     URL::URL url(href());
 
     // 3. If url or url's host is null, return the empty string.
-    if (url.host().has<Empty>())
+    // FIXME: How can url be null here?
```
Comment (shannonbooth) [inline Libraries/LibWeb/HTML/HTMLHyperlinkElementUtils.cpp:209]: The reason behind here is that the implementation doesn't follow spec - see: `HTMLHyperlinkElementUtils::hostname` as an example of how this should be done. `m_url` may be null.

File: Libraries/LibURL/Host.cpp [hunk 3]
```diff
@@ -152,4 +153,58 @@ String Host::serialize() const
...
+    auto trailing_dot = host_string.ends_with('.') ? "."sv : ""sv;
+
+    // 3. Let registrableDomain be the registrable domain determined by running the Public Suffix List algorithm with host as domain. [PSL]
+    auto registrable_domain = get_public_suffix(host_string);
+
+    // NOTE: get_public_suffix() returns Optional, but this algorithm assumes a value. Is that OK?
+    VERIFY(registrable_domain.has_value());
+
+    // 4. Assert: registrableDomain is an ASCII string that does not end with ".".
+    // FIXME: ASCII string
```
Comment (shannonbooth) [inline Libraries/LibURL/Host.cpp:203]: Maybe:

```cpp
    VERIFY(all_of(registrable_domain, is_ascii));
```

File: Libraries/LibWeb/HTML/WorkerLocation.cpp (hunk 1 above)
Comment (AtkinsSJ) [reply Libraries/LibWeb/HTML/WorkerLocation.cpp]: Done! I very nearly got caught in a deep pile of yaks with this. 😆 Managed to make it out alive after only a couple of changes.

File: Libraries/LibWeb/HTML/HTMLHyperlinkElementUtils.cpp (hunk 2 above)
Comment (AtkinsSJ) [reply Libraries/LibWeb/HTML/HTMLHyperlinkElementUtils.cpp]: Thanks, done that in a later commit.

**Comment by AtkinsSJ** [comment]
Changes:
- Implement the "is an ASCII string" checks
- Make a few methods infallible (as extra commits)
- Bring HTMLHyperlinkElementUtils::hostname() closer to spec (as an extra commit)

**Review by shannonbooth — APPROVED** [review]
lgtm bar the clang-format thing

**Comment by AtkinsSJ** [comment]
> lgtm bar the clang-format thing

Oops, that's a pain... looks like I have clang-format v19 installed by accident. 😅

**Pull request merged**